File: internal/commands/start.go
package commands

import (
	"fmt"
	"omnillm/internal/server"

	"github.com/spf13/cobra"
)

var StartCmd = &cobra.Command{
	Use:   "start",
	Short: "Start the LLM proxy server",
	Long: `Start the OmniLLM proxy server.

The server listens on --host:--port and exposes:
  - OpenAI-compatible API at /v1/chat/completions, /v1/models, /v1/embeddings
  - Anthropic-compatible API at /v1/messages
  - Admin API and UI at /admin/

Configuration precedence (highest to lowest):
  1. CLI flags
  2. Environment variables (OMNILLM_SERVER, OMNILLM_API_KEY, etc.)
  3. Files in ~/.config/omnillm/

The inbound --api-key defaults to a generated key stored in ~/.config/omnillm/api-key.`,
	Example: `  # Start with defaults (port 5000, github-copilot provider)
  omnillm start

  # Start on a different port with a specific provider
  omnillm start --port 8080 --provider openai-compatible

  # Start with an explicit API key and verbose logging
  omnillm start --api-key my-secret --verbose

  # Start with rate limiting (1 request per 3 seconds, wait instead of error)
  omnillm start --rate-limit 3 --wait

  # Print the Claude Code launch command after starting
  omnillm start --claude-code`,
	RunE: func(cmd *cobra.Command, args []string) error {
		port, err := cmd.Flags().GetInt("port")
		if err != nil {
			return fmt.Errorf("get port flag: %w", err)
		}
		host, err := cmd.Flags().GetString("host")
		if err != nil {
			return fmt.Errorf("get host flag: %w", err)
		}

		verbose, err := cmd.Flags().GetBool("verbose")
		if err != nil {
			return fmt.Errorf("get verbose flag: %w", err)
		}
		accountType, err := cmd.Flags().GetString("account-type")
		if err != nil {
			return fmt.Errorf("get account-type flag: %w", err)
		}
		manual, err := cmd.Flags().GetBool("manual")
		if err != nil {
			return fmt.Errorf("get manual flag: %w", err)
		}
		rateLimitSeconds, err := cmd.Flags().GetInt("rate-limit")
		if err != nil {
			return fmt.Errorf("get rate-limit flag: %w", err)
		}
		wait, err := cmd.Flags().GetBool("wait")
		if err != nil {
			return fmt.Errorf("get wait flag: %w", err)
		}
		githubToken, err := cmd.Flags().GetString("github-token")
		if err != nil {
			return fmt.Errorf("get github-token flag: %w", err)
		}
		claudeCode, err := cmd.Flags().GetBool("claude-code")
		if err != nil {
			return fmt.Errorf("get claude-code flag: %w", err)
		}
		console, err := cmd.Flags().GetBool("console")
		if err != nil {
			return fmt.Errorf("get console flag: %w", err)
		}
		showToken, err := cmd.Flags().GetBool("show-token")
		if err != nil {
			return fmt.Errorf("get show-token flag: %w", err)
		}
		proxyEnv, err := cmd.Flags().GetBool("proxy-env")
		if err != nil {
			return fmt.Errorf("get proxy-env flag: %w", err)
		}
		apiKey, err := cmd.Flags().GetString("api-key")
		if err != nil {
			return fmt.Errorf("get api-key flag: %w", err)
		}
		provider, err := cmd.Flags().GetString("provider")
		if err != nil {
			return fmt.Errorf("get provider flag: %w", err)
		}
		allowLocalEndpoints, err := cmd.Flags().GetBool("allow-local-endpoints")
		if err != nil {
			return fmt.Errorf("get allow-local-endpoints flag: %w", err)
		}
		enableConfigEdit, err := cmd.Flags().GetBool("enable-config-edit")
		if err != nil {
			return fmt.Errorf("get enable-config-edit flag: %w", err)
		}
		allowedChromeExtensions, err := cmd.Flags().GetStringSlice("allow-chrome-extension")
		if err != nil {
			return fmt.Errorf("get allow-chrome-extension flag: %w", err)
		}

		var rateLimit *int
		if cmd.Flags().Changed("rate-limit") {
			rateLimit = &rateLimitSeconds
		}

		options := server.StartOptions{
			Port:                     port,
			Host:                     host,
			Verbose:                  verbose,
			AccountType:              accountType,
			Manual:                   manual,
			RateLimit:                rateLimit,
			RateLimitWait:            wait,
			GithubToken:              githubToken,
			ClaudeCode:               claudeCode,
			Console:                  console,
			ShowToken:                showToken,
			ProxyEnv:                 proxyEnv,
			Provider:                 provider,
			APIKey:                   apiKey,
			AllowLocalEndpoints:      allowLocalEndpoints,
			EnableConfigEdit:         enableConfigEdit,
			AllowedChromeExtensionIDs: allowedChromeExtensions,
		}

		return server.RunServer(options)
	},
}

func init() {
	StartCmd.Flags().IntP("port", "p", 5000, "Port to listen on")
	StartCmd.Flags().String("host", "127.0.0.1", "IP or hostname to bind the server to")
	StartCmd.Flags().BoolP("verbose", "v", false, "Enable verbose logging")
	StartCmd.Flags().StringP("account-type", "a", "individual", "Account type to use (individual, business, enterprise)")
	StartCmd.Flags().Bool("manual", false, "Enable manual request approval")
	StartCmd.Flags().IntP("rate-limit", "r", 0, "Rate limit in seconds between requests")
	StartCmd.Flags().BoolP("wait", "w", false, "Wait instead of error when rate limit is hit")
	StartCmd.Flags().StringP("github-token", "g", "", "Provide GitHub token directly")
	StartCmd.Flags().BoolP("claude-code", "c", false, "Generate a command to launch Claude Code with proxy config")
	StartCmd.Flags().Bool("console", false, "Automatically open the admin console in your default browser")
	StartCmd.Flags().Bool("show-token", false, "Show tokens on fetch and refresh")
	StartCmd.Flags().Bool("proxy-env", false, "Initialize proxy from environment variables")
	StartCmd.Flags().String("provider", "github-copilot", "Provider to use (github-copilot, antigravity, alibaba, etc.)")
	StartCmd.Flags().String("api-key", "", "Inbound API key for protecting server routes")
	StartCmd.Flags().Bool("allow-local-endpoints", false, "Allow localhost/private OpenAI-compatible endpoints")
	StartCmd.Flags().Bool("enable-config-edit", false, "Allow editing external config files via admin API")
	StartCmd.Flags().StringSlice("allow-chrome-extension", nil, "Allow specific Chrome extension IDs for CORS (repeat flag or pass comma-separated IDs)")
}

File: src/start.ts
#!/usr/bin/env node

import { defineCommand } from "citty"
import clipboard from "clipboardy"
import consola from "consola"
import { serve } from "srvx"

import { loadConfig } from "./lib/config-db"
import { setupFileLogging, setLogLevel } from "./lib/logging"
import { selectModelWithFilter } from "./lib/model-selection"
import { ensurePaths } from "./lib/paths"
import { initProxyFromEnv } from "./lib/proxy"
import { generateEnvScript } from "./lib/shell"
import { state } from "./lib/state"
import { cacheVSCodeVersion } from "./lib/utils"
import { server } from "./server"

interface RunServerOptions {
  port: number
  verbose: boolean
  accountType: string
  manual: boolean
  rateLimit?: number
  rateLimitWait: boolean
  githubToken?: string
  claudeCode: boolean
  console: boolean
  showToken: boolean
  proxyEnv: boolean
  provider: string
}

/**
 * Initialize providers with auto-discovery but don't auto-activate them
 */
async function initializeProviders(
  _selectedProviderType: string,
): Promise<void> {
  // Load config (including saved provider instances) before discovery
  const { loadConfig } = await import("./lib/config-db")
  await loadConfig()

  // Import auto-discovery function
  const { discoverProviderInstances } = await import("./auth")

  // Auto-discover existing provider instances from providers folder
  await discoverProviderInstances()

  // No need to register default placeholders - users can add providers via the UI

  // Active providers are restored from config via loadConfig()
  consola.info("Providers initialized. Activation states restored from config.")
}

/**
 * Handle Claude Code setup and command generation
 */
async function handleClaudeCodeSetup(port: number): Promise<void> {
  // For Claude Code mode, we need models to be available
  // Check if we have any active providers and try to load models
  const { providerRegistry } = await import("./providers/registry")
  const activeProviders = providerRegistry.getActiveProviders()

  if (activeProviders.length === 0) {
    consola.warn(
      "No active providers found. You can add and activate providers through the admin UI at http://localhost:PORT/admin",
    )
    consola.info(
      "The proxy will still start, but you'll need to configure providers before making requests.",
    )
    return // Don't exit, just continue without models
  }

  // Try to load models from the first active provider
  const provider = activeProviders[0]
  try {
    consola.info(
      `Loading models from ${provider.name} for Claude Code setup...`,
    )
    const providerModels = await provider.getModels()
    state.models = providerModels
    consola.info(`Loaded ${providerModels.data.length} models`)
  } catch (error) {
    consola.error(`Failed to load models from ${provider.name}:`, error)
    consola.error("Please ensure the provider is properly authenticated.")
    process.exit(1)
  }

  const selectedModel = await selectModelWithFilter(
    state.models.data.map((model) => ({ id: model.id, name: model.name })),
    "Select a model to use with Claude Code",
  )

  const selectedSmallModel = await selectModelWithFilter(
    state.models.data.map((model) => ({ id: model.id, name: model.name })),
    "Select a small model to use with Claude Code",
  )

  const serverUrl = `http://localhost:${port}`
  const command = generateEnvScript(
    {
      ANTHROPIC_BASE_URL: serverUrl,
      ANTHROPIC_AUTH_TOKEN: "dummy",
      ANTHROPIC_MODEL: selectedModel,
      ANTHROPIC_DEFAULT_SONNET_MODEL: selectedModel,
      ANTHROPIC_SMALL_FAST_MODEL: selectedSmallModel,
      ANTHROPIC_DEFAULT_HAIKU_MODEL: selectedSmallModel,
      DISABLE_NON_ESSENTIAL_MODEL_CALLS: "1",
      CLAUDE_CODE_DISABLE_NONESSENTIAL_TRAFFIC: "1",
    },
    "claude",
  )

  try {
    clipboard.writeSync(command)
    consola.success("Copied Claude Code command to clipboard!")
  } catch {
    consola.warn("Failed to copy to clipboard")
  }

  consola.info(`\n${command}`)
}

/**
 * Handle interactive console mode
 */
async function handleConsoleMode(port: number): Promise<void> {
  const readline = await import("node:readline/promises")
  const stdin = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  })

  const serverUrl = `http://localhost:${port}`
  consola.info(`\nInteractive mode. ${serverUrl}/chat/completions`)
  consola.info('Type "exit" to quit.\n')

  while (true) {
    const prompt = await stdin.question("Prompt: ")
    if (prompt.trim().toLowerCase() === "exit") break

    try {
      const response = await fetch(`${serverUrl}/chat/completions`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          model: state.models?.data[0]?.id || "unknown",
          messages: [{ role: "user", content: prompt }],
          stream: false,
        }),
      })

      const data = (await response.json()) as {
        choices?: Array<{ message?: { content?: string } }>
      }
      const content = data.choices?.[0]?.message?.content
      if (content) {
        consola.info(`Response: ${content}\n`)
      } else {
        consola.error("No response content")
      }
    } catch (err) {
      const errorMessage = err instanceof Error ? err.message : "Unknown error"
      consola.error("Error:", errorMessage)
    }
  }

  stdin.close()
}
export async function runServer(options: RunServerOptions): Promise<void> {
  // Set up file logging first
  setupFileLogging()

  if (options.proxyEnv) {
    initProxyFromEnv()
  }

  if (options.verbose) {
    consola.level = 5
    consola.info("Verbose logging enabled")
  }

  state.accountType = options.accountType
  if (options.accountType !== "individual") {
    consola.info(`Using ${options.accountType} plan GitHub account`)
  }

  state.manualApprove = options.manual
  state.rateLimitSeconds = options.rateLimit
  state.rateLimitWait = options.rateLimitWait
  state.showToken = options.showToken

  await ensurePaths()
  await loadConfig()

  // Apply persisted log level now that DB is available
  try {
    const { ConfigStore } = await import("./lib/database")
    const saved = ConfigStore.get("log_level")
    if (saved !== null) {
      setLogLevel(Number.parseInt(saved))
    }
  } catch {
    // ignore if DB unavailable
  }

  // Initialize providers (discovery and registration only)
  await initializeProviders(options.provider)

  // Cache VSCode version if needed (doesn't require auth)
  if (options.provider === "github-copilot") {
    await cacheVSCodeVersion()
  }

  // That's it! No authentication, no model fetching during startup
  // All of that happens when users click "Activate" in the UI

  if (options.claudeCode) {
    await handleClaudeCodeSetup(options.port)
    return
  }

  if (options.console) {
    await handleConsoleMode(options.port)
    return
  }

  const serverUrl = `http://localhost:${options.port}`
  const adminUrl = `${serverUrl}/admin`
  state.port = options.port

  serve({
    fetch: server.fetch,
    port: options.port,
    onListen: () => {
      state.port = options.port
      consola.info(`🚀 Server running at ${serverUrl}`)
      consola.info(`📱 Admin UI at ${adminUrl}`)
    },
  })
}

export const start = defineCommand({
  meta: {
    name: "start",
    description: "Start the LLM proxy server",
  },
  args: {
    port: {
      alias: "p",
      type: "string",
      default: "4141",
      description: "Port to listen on",
    },
    verbose: {
      alias: "v",
      type: "boolean",
      default: false,
      description: "Enable verbose logging",
    },
    "account-type": {
      alias: "a",
      type: "string",
      default: "individual",
      description: "Account type to use (individual, business, enterprise)",
    },
    manual: {
      type: "boolean",
      default: false,
      description: "Enable manual request approval",
    },
    "rate-limit": {
      alias: "r",
      type: "string",
      description: "Rate limit in seconds between requests",
    },
    wait: {
      alias: "w",
      type: "boolean",
      default: false,
      description:
        "Wait instead of error when rate limit is hit. Has no effect if rate limit is not set",
    },
    "github-token": {
      alias: "g",
      type: "string",
      description:
        "Provide GitHub token directly (must be generated using the `auth` subcommand)",
    },
    "claude-code": {
      alias: "c",
      type: "boolean",
      default: false,
      description: "Generate a command to launch Claude Code with proxy config",
    },
    console: {
      type: "boolean",
      default: false,
      description:
        "Automatically open the admin console in your default browser",
    },
    "show-token": {
      type: "boolean",
      default: false,
      description: "Show tokens on fetch and refresh",
    },
    "proxy-env": {
      type: "boolean",
      default: false,
      description: "Initialize proxy from environment variables",
    },
    provider: {
      type: "string",
      default: "github-copilot",
      description:
        "Provider to use (github-copilot, antigravity, alibaba, etc.)",
    },
  },
  run({ args }) {
    const rateLimitRaw = args["rate-limit"]
    const rateLimit =
      rateLimitRaw ? Number.parseInt(rateLimitRaw, 10) : undefined

    return runServer({
      port: Number.parseInt(args.port, 10),
      verbose: args.verbose,
      accountType: args["account-type"],
      manual: args.manual,
      rateLimit,
      rateLimitWait: args.wait,
      githubToken: args["github-token"],
      claudeCode: args["claude-code"],
      console: args.console,
      showToken: args["show-token"],
      proxyEnv: args["proxy-env"],
      provider: args.provider,
    })
  },
})
